hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OmLockOpr.java

```diff
@@ -0,0 +1,268 @@
+    List<Lock> locks = omLockInfo.getLocks();
+    long startTime = Time.monotonicNowNanos();
+    locks.add(lockFunction.apply(name));
+    long endTime = Time.monotonicNowNanos();
+    omLockInfo.add(endTime - startTime, type);
+    omLockInfo.setLockTakenTime(endTime);
+    lockMonitorMap.put(omLockInfo, omLockInfo);
+    return omLockInfo;
+  }
+
+  public OmLockInfo obsLock(String bucketName, String keyName) throws IOException {
+    OmLockInfo omLockInfo = new OmLockInfo();
+    List<Lock> locks = omLockInfo.getLocks();
+    // bucket read lock
+    long startTime = Time.monotonicNowNanos();
+    locks.add(bucketLocking.readLock(bucketName));
+    try {
+      // key lock with bucket uniqueness as same key can be present across bucket
+      locks.add(keyLocking.lock(bucketName + "/" + keyName));
+      long endTime = Time.monotonicNowNanos();
+      omLockInfo.add(endTime - startTime, OmLockInfo.LockOpType.WAIT);
+      omLockInfo.setLockTakenTime(endTime);
+      lockMonitorMap.put(omLockInfo, omLockInfo);
+      return omLockInfo;
+    } catch (IOException ex) {
+      writeUnlock(omLockInfo);
+      throw ex;
+    }
+  }
+
+  public OmLockInfo obsLock(String bucketName, List<String> keyList) throws IOException {
+    OmLockInfo omLockInfo = new OmLockInfo();
+    List<Lock> locks = omLockInfo.getLocks();
+    // bucket read lock
+    long startTime = Time.monotonicNowNanos();
+    locks.add(bucketLocking.readLock(bucketName));
+    try {
+      // key lock with bucket uniqueness as same key can be present across bucket
+      List<String> bucketKeyList = new ArrayList<>();
+      keyList.forEach(e -> bucketKeyList.add(bucketName + "/" + e));
+      locks.addAll(keyLocking.lock(bucketKeyList));
+      long endTime = Time.monotonicNowNanos();
+      omLockInfo.add(endTime - startTime, OmLockInfo.LockOpType.WAIT);
+      omLockInfo.setLockTakenTime(endTime);
+      lockMonitorMap.put(omLockInfo, omLockInfo);
+      return omLockInfo;
+    } catch (IOException ex) {
+      writeUnlock(omLockInfo);
+      throw ex;
+    }
+  }
+
+  public void writeUnlock(OmLockInfo lockInfo) {
+    unlock(lockInfo, OmLockInfo.LockOpType.WRITE);
+  }
+
+  public void readUnlock(OmLockInfo lockInfo) {
+    unlock(lockInfo, OmLockInfo.LockOpType.READ);
+  }
+
+  private void unlock(OmLockInfo lockInfo, OmLockInfo.LockOpType type) {
+    Collections.reverse(lockInfo.getLocks());
+    lockInfo.getLocks().forEach(Lock::unlock);
+    if (lockInfo.getLockTakenTime() > 0) {
+      lockInfo.add(Time.monotonicNowNanos() - lockInfo.getLockTakenTime(), type);
+    }
+    lockInfo.getLocks().clear();
+    lockMonitorMap.remove(lockInfo);
+  }
+
+  public void monitor() {
+    long curTime = Time.monotonicNowNanos() - MONITOR_LOCK_THRESHOLD_NS;
+    for (OmLockInfo lockInfo : lockMonitorMap.values()) {
+      if ((curTime - lockInfo.getLockTakenTime()) > 0) {
+        LOG.warn("Lock {} is crossing threshold {}: ", lockInfo, MONITOR_LOCK_THRESHOLD_NS);
```

Comment:
This will completely flood the logs in a deadlock situation, regardless of the interval the check runs. We need some in-memory tracker for which locks have already had warnings issued.